### DISCUSSION TITLE
Could ref field not be marshal as pointer?

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/runtime#78316] [Q&A]
Test.cpp
```
struct T
{
	int* pNum;
};

extern "C" __declspec(dllexport) void Test(T t) {
	int* pNum = t.pNum;
	cout << pNum << endl;
}
```

Test.cs
```
public static void Main(string[] args)
{
    int[] nums = GC.AllocateArray<int>(10, true);
    T t = new T(ref nums[0]);

    Test(t);
}

[DllImport("Test.cpp")]
public static extern void Test(T t);

public ref struct T
{
    public T(ref int pNum)
    {
        this.pNum = ref pNum;
    }

    public ref int pNum;
}
```

This code, why does the console always output ```0000000000000040``` instead of a pointer to ```nums[0]```? 

```
public static void Main(string[] args)
{
    int[] nums = GC.AllocateArray<int>(10, true);

    Test(ref nums[0]);
}

[DllImport("Test.cpp")]
public static extern void Test(ref int t);
```
But this code, the console always output a pointer to nums[0].

Why does the ```ref int t``` be marshal as a pointer to ```nums[0]```, but the ```T t``` is not?

**Comment by hez2010** [comment]
Maybe you can try using CustomMarshaller to provide your own Marshaller.

**Comment by NN---** [comment]
Can you use ref T in Import to pass it by reference instead of passing by value ?

```cs
[DllImport("Test.cpp")]
public static extern void Test(ref T t);
```

**Reply by ghost** [reply]
No, this will pass a pointer to ```T t```, not ```nums[0]```.

**Reply by NN---** [reply]
Got you. You say `ref int` is not marshaled as pointer to the object it references to.
Do you think it is an expected behavior?

**Reply by ghost** [reply]
I just don't understand why could the ```ref int t``` be marshal as a pointer to nums[0], but the ```T t``` is marshal as ```0000000000000040```.

**Reply by hez2010** [reply]
The interop recommendation is to use blittable types for interop. You may use `Unsafe.AsPointer` to cast the ref to pointer.

**Reply by vitek-karas** [reply]
@elinor-fung @AaronRobinsonMSFT for interop expertise.

**Reply by tannergooding** [reply]
> You may use Unsafe.AsPointer to cast the ref to pointer.

This can leave a GC hole. You should pin the `ref` using a `fixed` statement and pass that pointer instead.
